pkg/server/server.go

```diff
@@ -491,6 +498,22 @@ func (s *Server) Run(ctx context.Context) error {
 		return err
 	}
 
+	// add post start hook that starts all registered controllers.
+	s.AddPostStartHook("kcp-start-controllers", func(hookContext genericapiserver.PostStartHookContext) error {
+		logger := klog.FromContext(ctx).WithValues("postStartHook", "kcp-start-controllers")
+
+		if s.ExtraConfig.LeaderElectionEnabled {
+			go s.leaderElectAndRun(goContext(hookContext))
+		} else {
+			for controllerName, controller := range s.controllers {
+				logger := logger.WithValues("controller", controllerName)
+				go s.runController(goContext(hookContext), controller, logger)
```

Comment:
nit: this bit of logic to start registered controllers can probably be put into a function `startControllers` so that we can abstract away anything else needed like recovery handlers etc. that might be needed to add in later on.

Reply:
I moved this to `s.startControllers`.